mysql-plugin/src/test/java/io/cdap/plugin/mysql/MysqlSinkTestRun.java

```diff
@@ -104,12 +103,13 @@ public void testDBSinkWithInvalidFieldLogicalType() throws Exception {
 
   @Test
   public void testDBSinkWithDBSchemaAndInvalidData() throws Exception {
-    String stringColumnName = "NAME";
-    startPipelineAndWriteInvalidData(stringColumnName, getSinkConfig(), DATAPIPELINE_ARTIFACT);
+    // MySQL JDBC connector allows to write integer values to STRING column. Use ENUM column instead.
+    String enumColumnName = "ENUM_COL";
+    startPipelineAndWriteInvalidData(enumColumnName, getSinkConfig(), DATAPIPELINE_ARTIFACT);
```

Comment:
In the scope of this PR, `AbstractDBSink#transform` [method was modified](https://github.com/data-integrations/database-plugins/pull/46/files#diff-5570db2ea5dd53e68f60ef41d9c8f6e8L211) to emit original input record and not create a new one(based on the output schema). 
```
  @Override
  public void transform(StructuredRecord input, Emitter<KeyValue<DBRecord, NullWritable>> emitter) {
    emitter.emit(new KeyValue<>(getDBRecord(input), null));
  }
```

This allows us to provide the following schema compatibility:
1. Schema.Type.INT -> Schema.LogicalType.DECIMAL (if precision of actual decimal logical type >= 10)
2. Schema.Type.LONG -> Schema.LogicalType.DECIMAL (if precision of actual decimal logical type >= 19)
3. Schema.Type.FLOAT -> Schema.LogicalType.DECIMAL (primitive value can be rounded to match actual schema)
4. Schema.Type.DOUBLE -> Schema.LogicalType.DECIMAL (primitive value can be rounded to match actual schema)

These changes broke `testDBSinkWithDBSchemaAndInvalidData` test, since now we are trying to write original input record(and not transformed one) and will fail only if JDBC connector does not allow to perform the write. 

MySQL JDBC connector allows to write integer values to STRING column, converting to string implicitly, so here we use `ENUM` column instead.

Another approach would be revert `AbstractDBSink#transform` changes and transform primitives into corresponding instances of `BigDecimal` explicitly.

Reply:
> I'm confused how that has to do with data types. 

From your discussion with Maksym at the meeting on June 25, I heard that we should not transform a field, instead, we should fail writing field if it's not compatible.
Seems that I got it wrong.

We can support writing primitives to decimal columns in two ways:
1. Transform primitives into corresponding instances of BigDecimal explicitly to match the output schema

Handle the case when output schema expects decimal logical type but we got valid primitive. Transform primitives into corresponding instances of BigDecimal.

2. Emit original record and write primitives using corresponding methods of PreparedStatement

For example, it's possible to write `int` to `NUMBER(38, 0)` column using `PreparedStatement#setInt` method.

Drawbacks: 
We lose some sort of validation. The broken test is a good example.
With the **old implementation**, if we create a new record based on the output schema, we will **fail writing an integer to string** column in MySQL Sink.
However, the MySQL JDBC connector allows doing so and **proposed implementation will store integer to string column.**  

* Also, we can map Oracle NUMBER types as I mentioned in the https://github.com/data-integrations/database-plugins/pull/46#issuecomment-511931265
and do not support writing primitives to decimal columns at all.

*******************************************************************

> Previously, the record was being created in order to filter out fields that are present in the input schema but not in the output table.

Proposed implementation [assumes that some of the output schema fields can be absent in the actual record](https://github.com/data-integrations/database-plugins/pull/46/files#diff-4799ee249e76f7b18abb28c4d52191b7R270):

```
  protected void writeToDB(PreparedStatement stmt, @Nullable Schema.Field field, int fieldIndex) throws SQLException {

    int sqlIndex = fieldIndex + 1;
    int sqlType = columnTypes.get(fieldIndex).getType();
    if (field == null) {
      // Some of the fields can be absent in the record
      stmt.setNull(sqlIndex, sqlType);
      return;
    }

   ...
```

I would like to show you the differences in the behavior of proposed and old implementation.
Let's consider the following scenario with simple source-to-sink pipeline. Tested this with Oracle plugins built from this branch and from 5f2118190047afe0828794fc52e91716de5080bb.

Source table created as: 
```
create table source(FIRST VARCHAR(50), SECOND VARCHAR(50), THIRD VARCHAR(50));
insert into source values('first', 'second', 'third');
```

Sink table:
```
create table sink(FIRST VARCHAR(50));
```

1. No schema specified in UI(nor manually, nor via 'Get Schema' button)

**Proposed implementation:** Succeeded;

**Old implementation:**
Deployed successfully, failed with:
```
java.lang.IllegalArgumentException: Input field 'SECOND' is not found in columns
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:119) ~[com.google.guava.guava-13.0.1.jar:na]
	at io.cdap.plugin.db.batch.sink.AbstractDBSink.transform(AbstractDBSink.java:209) ~[database-commons-1.3.0-SNAPSHOT.jar:na] 

```

2. Output schema with single `FIRST` field specified 

**Proposed implementation:** Succeeded;

**Old implementation:** Succeeded;

3. Output schema with all fields specified(`FIRST`, `SECOND`, `THIRD`)

**Proposed implementation:** 
Failed to deploy with:
```
Failed to configure pipeline: Error configuring stage 'OracleSink': java.sql.SQLException: Invalid column name
```

**Old implementation:**
Deployed successfully, failed with:
```
java.sql.SQLException: Invalid column name
	at oracle.jdbc.driver.OracleStatement.getColumnIndex(OracleStatement.java:3757) ~[na:na]
	at oracle.jdbc.driver.InsensitiveScrollableResultSet.findColumn(InsensitiveScrollableResultSet.java:264) ~[na:na]
```

*********************************************************************

Could you please share your opinion regarding the following questions:
1. What is preferred mapping for Oracle NUMBER types?
2. Should we support writing primitives to decimal columns? If so, what approach is preferred? Should I revert `AbstractDBSink#transform` method changes?

Thanks in advance!

Reply:
>  For example, if the input field is a CDAP int, we should not write it to a database varchar column.

Currently, there is no validation for each received input record. Only explicit schema is validated.
We just copying corresponding fields from input record to a new one, created based on the output schema. For instance, with the old implementation, we will be able to write values of decimal logical type to BINARY columns(if there is no schema set, so no validation is triggered). 

> Or are you saying the actual values there didn't match what is in the schema?

Yes. 
Let's consider the following case:

0. Having old implementation of [AbstractDbSink#transform](https://github.com/data-integrations/database-plugins/blob/develop/database-commons/src/main/java/io/cdap/plugin/db/batch/sink/AbstractDBSink.java#L211) method:
```
  @Override
  public void transform(StructuredRecord input, Emitter<KeyValue<DBRecord, NullWritable>> emitter) {
    // Create StructuredRecord that only has the columns in this.columns
    List<Schema.Field> outputFields = new ArrayList<>();
    for (Schema.Field field : input.getSchema().getFields()) {
      Preconditions.checkArgument(columns.contains(field.getName()), "Input field '%s' is not found in columns",
                                  field.getName());
      outputFields.add(field);
    }
    StructuredRecord.Builder output = StructuredRecord.builder(outputSchema);
    for (String column : columns) {
      output.set(column, input.get(column));
    }

    emitter.emit(new KeyValue<>(getDBRecord(output), null));
  }
```

1. Sink table created as:

```
create table sink(INT_COL INTEGER, SMALL_COL SMALLINT); 
```

2. There is a [logic in the Sink plugin, that infers the schema](https://github.com/data-integrations/database-plugins/blob/develop/database-commons/src/main/java/io/cdap/plugin/db/batch/sink/AbstractDBSink.java#L185) if it's not set. 

```
    try {
      if (Objects.nonNull(outputSchema)) {
        validateSchema(driverClass, dbSinkConfig.tableName, outputSchema);
      } else {
        outputSchema = inferSchema(driverClass);
      }
    } finally {
      DBUtils.cleanup(driverClass);
    }
```
This schema will be used in `AbstractDbSink#transform` method to build an actual record.
The inferred schema will contain two fields of decimal logical type.

_I'm wondering is it required to infer a schema in the Sink plugin? Isn't it passed by the Source? Or there are some cases when this will be useful?_ 

3. Transform receives input record with primitives

Input record's schema contains two fields with the same names but of `Type.INT` instead of decimal logical type.

4. `AbstractDbSink#transform` method creates a new `StructuredRecord` with decimal schema, just copying corresponding values.

Here we have the situation when actual values don't match the schema of the newly created `StructuredRecord`.

5. We fail to access primitives as values of decimal logical type while writing

We expect to access values via `StructuredRecord#getDecimal` since according to the record's schema they are of decimal logical type.

*************************************************************************************************
So i see the following options:
* Do not infer a schema in the Sink plugin
* Do not support writing primitives with inferred Sink schema
* Transform primitives of the `inputRecord` into corresponding instances of `BigDecimal`
* Do not transform original `inputRecord`, write it using its schema. So, we will write field of the `inputRecord` of type `INT` as integer(`PreparedStatement#setInt`), and will fail if JDBC connector does not allow us doing so. This is how the proposed implementation works now. Should i change it to on of the options above?

Reply:
OK I think I understand the problem better now.

I must not have looked carefully when inferSchema() was first added. It is never correct to guess the input schema based on the output database table. The input schema is determined by the previous stage in the pipeline -- not by the output table. In the prepareRun() method, it is just configuring the output format so it uses inferSchema() to configure the column names. This is fine. In the initialize() method, it should *not* use inferSchema(). It is incorrect to assume the records have a schema that match the one coming from the db table.

When context.getInputSchema() is not null, it is guaranteed that every record given to the transform() method has that schema. When it is null, there is no guarantee. The schema can be anything. Incoming records could have different schemas.

What this plugin is responsible for is mapping the types from the incoming record into the type expected by the oracle jdbc driver. Does the driver require setting a BigDecimal for 'integer' columns? Then the plugin should ensure that incoming ints are changed to BigDecimal before writing to the DB.

Reply:
>  In the initialize() method, it should not use inferSchema(). It is incorrect to assume the records have a schema that match the one coming from the db table.

[CDAP documentation says](https://docs.cask.co/cdap/current/en/developer-manual/pipelines/developing-plugins/creating-a-plugin.html#H2231) that `initialize()` method must create any resources required by `transform()` method. 
I believe that's why `inferSchema()` was added to this method.

It configures the column names. These names are used by `setResultSetMetadata` method to initialize `columnTypes` list, which is actually used in the `transform()` method. 

The proposed implementation does not use the inferred schema for other cases except this one. 
The old implementation uses the inferred schema in `transfrom()` method to create a new record. 


> When it is null, there is no guarantee. The schema can be anything. Incoming records could have different schemas.

The proposed implementation works exactly in this way. It uses the schema of received record to write record's fields. 
That's why MySQL test, which expects that an attempt to write int to string column must fail, was broken.

The old implementation:
1. the test does not set a schema
2. we receive int field of Type.INT 
3. we copy it to a new record, using inferred schema, which infers that this field is of Type.STRING
4. an attempt to get the int from a StructuredRecord fails

Proposed implementation:
1. the test does not set a schema
2. we receive int field of Type.INT
3. an attempt to write int to string column succeeds since JDBC connector allows doing so

> Does the driver require setting a BigDecimal for 'integer' columns?

No, it allows us to set Java integers and that's how the proposed implementation writes primitives to 'integer' columns.